language/move-vm/integration-tests/src/tests/limits_tests.rs

```diff
@@ -0,0 +1,2788 @@
+// Copyright (c) The Move Contributors
+// SPDX-License-Identifier: Apache-2.0
+
+#![allow(dead_code)]
+#![allow(unused_must_use)]
+#![allow(unused_imports)]
+
+use move_binary_format::{
+    errors::VMResult,
+    file_format::{
+        AbilitySet, AddressIdentifierIndex, Bytecode, Bytecode::*, CodeUnit, CompiledModule,
+        Constant, ConstantPoolIndex, FieldDefinition, FunctionDefinition, FunctionHandle,
+        FunctionHandleIndex, FunctionInstantiation, FunctionInstantiationIndex, IdentifierIndex,
+        ModuleHandle, ModuleHandleIndex, Signature, SignatureIndex, SignatureToken,
+        SignatureToken::*, StructDefInstantiation, StructDefInstantiationIndex, StructDefinition,
+        StructDefinitionIndex, StructFieldInformation, StructHandle, StructHandleIndex,
+        StructTypeParameter, TypeSignature, Visibility::*,
+    },
+};
+use move_bytecode_verifier::{verify_module, VerifierConfig};
+use move_core_types::{
+    account_address::AccountAddress,
+    identifier::{IdentStr, Identifier},
+    language_storage::{ModuleId, StructTag, TypeTag},
+    vm_status::StatusCode,
+};
+use move_vm_runtime::{
+    config::VMConfig,
+    move_vm::MoveVM,
+    session::{SerializedReturnValues, Session},
+};
+use move_vm_test_utils::{
+    gas_schedule::{Gas, GasStatus, INITIAL_COST_SCHEDULE},
+    InMemoryStorage,
+};
+use move_vm_types::loaded_data::runtime_types::Type;
+use once_cell::sync::Lazy;
+use std::time::Instant;
+
+const MODULE_NAME: &str = "Mod";
+const OUTER_NAME: &str = "Outer";
+const INNER_NAME: &str = "Inner";
+const FIELD_NAME: &str = "inner";
+const ENTRY_POINT_NAME_1: &str = "entry_point";
+const ENTRY_POINT_NAME_2: &str = "entry_point_one_ty_arg";
+const ENTRY_POINT_NAME_3: &str = "entry_point_mul_ty_args";
+
+const RECURSIVE_NAME: &str = "recursive";
+const EMPTY_NAME: &str = "empty";
+
+static DEFAULT_SIGNATURES: Lazy<Vec<Signature>> =
+    Lazy::new(|| vec![Signature(vec![]), Signature(vec![TypeParameter(0)])]);
+
+#[test]
+fn test_limit_vector() {
+    let verifier = VerifierConfig {
+        max_type_instantiation_size: Some(128),
+        ..Default::default()
+    };
+    // instantiated tests
+    let res = run_with_module(&verifier, vector_instantiated_2_nodes);
+    res.expect("max_type_instantiation_size(128) vector_instantiated_2_nodes failed");
```

Comment:
I worry a bit about maintaining these tests. I also worry for auditing in that I find it hard to grok what tests are here. Thoughts on these issues? Do you share any concerns here? 

I think we could cook up a Rust macro (just a normy one, nothing procedural) for stamping out these cases. What would you think about that? Or do you have anything else in mind?

Reply:
yeah those tests are a workbench for someone that wanted to look into building some semi-automatic test generation around instantiation. Which ruled and rules out the macros for now.
In other words, I want a simple abstraction over the instructions that do instantiations but I did not know exactly what that would look like. So all those tests are an attempt to understand the different scenarios. And they are all still there because I am still thinking about them :) 
There are several ways they can be unified, still thinking about how to break it up. And allow generation of test case with proptest and/or fuzzying.
So super happy to have deep conversations about this if they move it forward faster. Otherwise wear your seatbelt and let's move on.
I was surprised that I could build so simple test case (in terms of instruction sequences) for complex instantiations, and I am still thinking if they cover all we want.
Macros would set things in stone too much and make you lose focus on the real problem you are trying to solve. They still may be there eventually.

Reply:
maybe it is also worth to point out they are testing all generic opcodes in different instantiations tests.
I know it does not help a lot, but I am hoping they are reasonably well covered and exposed real issues.

Reply:
and finally, the instantiation tests of the other PR is also something that will follow a final re-org of the instantiation testing code.

I think finally we should have something comprehensible to all